lib/datasource/maven/index.ts

```diff
@@ -166,74 +156,83 @@ async function createUrlForDependencyPom(
-    return null;
+    return releaseMap;
   }
 
-  const cacheNamespace = 'datasource-maven-metadata';
+  const cacheNs = 'datasource-maven:head-requests';
   const cacheKey = `${repoUrl}${dependency.dependencyUrl}`;
-  let artifactsInfo: ArtifactsInfo | null =
-    await packageCache.get<ArtifactsInfo>(cacheNamespace, cacheKey);
-
-  // If the cache contains any artifacts that we were previously unable to determine if they exist,
-  // retry the existence checks on them.
-  if (!isValidArtifactsInfo(artifactsInfo, versions)) {
-    // For each version, determine if there is a POM file available for it
-    const results: ArtifactInfoResult[] = await pMap(
-      versions,
-      async (version): Promise<ArtifactInfoResult | null> => {
-        // Create the URL that the POM file should be available at
-        const artifactUrl = getMavenUrl(
-          dependency,
-          repoUrl,
-          await createUrlForDependencyPom(version, dependency, repoUrl)
-        );
-
-        // Return an ArtifactInfoResult that maps the version to the result of the check if the POM file exists in the repo
-        return [version, await isHttpResourceExists(artifactUrl)];
-      },
-      { concurrency: 5 }
-    );
+  let workingReleaseMap: ReleaseMap = await packageCache.get<ReleaseMap>(
+    cacheNs,
+    cacheKey
+  );
 
-    artifactsInfo = results.reduce(
-      (acc, [key, value]) => ({
-        ...acc,
-        [key]: value,
-      }),
-      {}
-    );
+  if (!workingReleaseMap) {
+    workingReleaseMap = {};
 
-    // Retry earlier for status other than 404
-    const cacheTTL = Object.values(artifactsInfo).some((x) => x === null)
-      ? 60
-      : 24 * 60;
+    const unknownVersions = Object.entries(releaseMap)
+      .filter(([version, release]) => {
+        const isDiscoveredOutside = !!release;
+        const isDiscoveredInsideAndCached = !is.undefined(
+          workingReleaseMap[version]
+        );
+        const isDiscovered = isDiscoveredOutside || isDiscoveredInsideAndCached;
+        return !isDiscovered;
+      })
+      .map(([k]) => k);
+
+    if (unknownVersions.length) {
+      let retryEarlier = false;
+      const queue = unknownVersions.map(
+        (version) => async (): Promise<void> => {
+          const pomUrl = await createUrlForDependencyPom(
+            version,
+            dependency,
+            repoUrl
+          );
+          const artifactUrl = getMavenUrl(dependency, repoUrl, pomUrl);
+          const res = await isHttpResourceExists(artifactUrl);
+          const release: Release = { version };
+
+          if (is.string(res)) {
+            release.releaseTimestamp = res;
+          }
+
+          // Retry earlier for error status other than 404
+          if (res === null) {
+            retryEarlier = true;
+          }
+
+          workingReleaseMap[version] = res ? release : null;
+        }
+      );
+
+      await pAll(queue, { concurrency: 5 });
```

Comment:
not sure if this will cause too many concurrent connections?

Reply:
It meant to be only 5 at the particular moment of time. My intention was to preserve current behavior, but rewrite with imperative `pAll` that looks more readable (at least, to me personally).